pkg/metrics/metrics.go

```diff
@@ -0,0 +1,164 @@
+// Namespace and subsystem are applied to all produced metrics.
+func (mr *MetricRecorder) NewPrometheusProvider() *MetricRecorder {
+	return &MetricRecorder{
+		Logger:    mr.Logger,
+		Scheme:    mr.Scheme,
+		Version:   mr.Version,
+		Project:   mr.Project,
+		Cluster:   mr.Cluster,
+		Slice:     mr.Slice,
+		Namespace: mr.Namespace,
+		Subsystem: mr.Subsystem,
+	}
+}
+
+func (mr *MetricRecorder) RecordMetric(ctx context.Context, m *Metric) error {
+	switch m.Type {
+	case MetricTypeGauge:
+		mr.recordGauge(ctx, m)
+		break
+	case MetricTypeCounter:
+		mr.recordCounter(ctx, m)
+		break
+	case MetricTypeHistogram:
+		mr.recordHistogram(ctx, m)
+		break
+	case MetricTypeSummary:
+		mr.recordSummary(ctx, m)
+		break
+	default:
+		mr.Logger.Errorf("Unknown metric type: %v", m.Type)
+	}
+	return nil
+}
+
+func (mr *MetricRecorder) recordGauge(ctx context.Context, m *Metric) {
+	mr.Logger.Infof("Recording gauge metric: %v", m)
+	metric := prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: util.FlattenKey(mr.Namespace),
+			Subsystem: util.FlattenKey(mr.Subsystem),
+			Name:      util.FlattenKey(m.Name),
+			Help:      m.Help,
+		},
+		append(fixedLabels, util.KeysFromMap(m.Labels)...),
+	)
+	metric.With(
+		util.MergeMaps(
+			prometheus.Labels{
+				"slice_project": util.FlattenKey(mr.Project),
+				"slice_cluster": util.FlattenKey(mr.Cluster),
+				"slice_name":    util.FlattenKey(mr.Slice),
+			}, util.FlattenMap(m.Labels)),
+	).Set(m.Value)
+}
+
+func (mr *MetricRecorder) recordCounter(ctx context.Context, m *Metric) {
+	mr.Logger.Infof("Recording counter metric: %v", m)
+	metric := prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Namespace: util.FlattenKey(mr.Namespace),
+			Subsystem: util.FlattenKey(mr.Subsystem),
+			Name:      util.FlattenKey(m.Name),
+			Help:      m.Help,
+		},
+		append(fixedLabels, util.KeysFromMap(m.Labels)...),
+	)
+	metric.With(
+		util.MergeMaps(
+			prometheus.Labels{
+				"slice_project": util.FlattenKey(mr.Project),
+				"slice_cluster": util.FlattenKey(mr.Cluster),
+				"slice_name":    util.FlattenKey(mr.Slice),
+			}, util.FlattenMap(m.Labels)),
+	).Add(m.Value)
+}
+
+func (mr *MetricRecorder) recordHistogram(ctx context.Context, m *Metric) {
+	mr.Logger.Infof("Recording histogram metric: %v", m)
+	start := time.Now()
+	metric := prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Namespace: util.FlattenKey(mr.Namespace),
+			Subsystem: util.FlattenKey(mr.Subsystem),
+			Name:      util.FlattenKey(m.Name),
+			Help:      m.Help,
+			Buckets:   m.histogramBuckets,
+		},
+		append(fixedLabels, util.KeysFromMap(m.Labels)...),
+	)
+	metric.With(
+		util.MergeMaps(
+			prometheus.Labels{
+				"slice_project": util.FlattenKey(mr.Project),
+				"slice_cluster": util.FlattenKey(mr.Cluster),
+				"slice_name":    util.FlattenKey(mr.Slice),
+			}, util.FlattenMap(m.Labels)),
+	).Observe(time.Since(start).Seconds())
```

Comment:
Histogram should provide time as an input from user.  In this case, the value will be always near 0 since start is declared in the same method.

Reply:
Taken `Time` as a user input. And made correction as `.Observe(time.Since(m.Time).Seconds())`